modules/fancymeshrenderer/glsl/fancymeshrenderer.frag

```diff
@@ -0,0 +1,346 @@
+#endif
+#ifdef ALPHA_DENSITY
+        float density_alpha = alphaSettings.baseDensity / (frag.area * angle * 100);
+        alpha *= pow(min(1, density_alpha), alphaSettings.densityExp);
+#endif
+#ifdef ALPHA_SHAPE
+        float shape_alpha = 4 * frag.area / 
+            (sqrt(3) * max(max(frag.sideLengths.x * frag.sideLengths.y,
+                frag.sideLengths.x*frag.sideLengths.z),
+                frag.sideLengths.y*frag.sideLengths.z));
+        alpha *= pow(min(1, shape_alpha), alphaSettings.shapeExp);
+#endif
+    }
+    color.a *= alpha;
+
+    //==================================================
+    // EDGES
+    //==================================================
+#if defined(DRAW_EDGES) || defined(DRAW_SILHOUETTE)
+    {
+        //obtain silhouette flags
+#ifdef DRAW_SILHOUETTE
+        bvec3 silhouettes = frag.silhouettes;
+#else
+        bvec3 silhouettes = bvec3(false);
+#endif
+
+        //compute if we are on an edge or not
+#ifdef DRAW_EDGES_SMOOTHING
+        //smoothing
+        float isEdgeSmoothed = 1;
+        float isSilhouettesSmoothed = 1;
+        vec3 dx = dFdxFinest(frag.edgeCoordinates);
+        vec3 dy = dFdyFinest(frag.edgeCoordinates);
+        for (int i=0; i<3; ++i) {
+            //Distance to the line
+            float d = abs(frag.edgeCoordinates[i]-1) / length(vec2(dx[i], dy[i]));
+            float fraction = frag.edgeCoordinates[i]<1 ? (1-(0.5*d + 0.5)) : (0.5*d+0.5);
+            isEdgeSmoothed *= 1 - clamp(fraction, 0, 1);
+            if (silhouettes[i]) isSilhouettesSmoothed *= 1 - clamp(fraction, 0, 1);
+        }
+        isEdgeSmoothed = 1 - isEdgeSmoothed;
+        isSilhouettesSmoothed = 1 - isSilhouettesSmoothed;
+#else
+        float isEdge = any(greaterThan(frag.edgeCoordinates,vec3(1))) ? 1.0f : 0.0f;
+        float isEdgeSmoothed = isEdge;
+        float isSilhouettesSmoothed = any(greaterThan(frag.edgeCoordinates,vec3(1)) && silhouettes) ? 1.0f : 0.0f;
+#endif
+        //blend in edge color
+        if (settings.showEdges) {
+            color.rgb = mix(color.rgb, settings.edgeColor.rgb, isEdgeSmoothed*min(1,settings.edgeColor.a));
+            color.a = mix(color.a, 1, isEdgeSmoothed*max(0, settings.edgeColor.a-1));
+        }
+#ifdef DRAW_SILHOUETTE
+        //blend in silhouette
+        color.rgb = mix(color.rgb, silhouetteColor.rgb, isSilhouettesSmoothed*min(1,silhouetteColor.a));
+        color.a = mix(color.a, 1, isSilhouettesSmoothed*max(0, silhouetteColor.a-1));
+#endif
+    }
+#endif
+
+    //==================================================
+    // HATCHING
+    //==================================================
+#ifdef SEND_TEX_COORD
+    vec2 texCoord = frag.texCoord;
+    float stripeStrength = 1;
+    if (settings.hatchingMode == 1 || settings.hatchingMode == 3 || settings.hatchingMode == 4) {
+        //hatch in u-direction
+        float lambdaS = length(vec2(dFdxFinest(texCoord.x), dFdyFinest(texCoord.x))) + 0.000000001;
```

Comment:
should be possible to extract this to a function, seem to be duplicated three times